src/resdk/collection_tables.py

```diff
@@ -0,0 +1,290 @@
+""".. Ignore pydocstyle D400.
+
+=================
+Collection Tables
+=================
+
+.. autoclass:: CollectionTables
+    :members:
+
+    .. automethod:: __init__
+"""
+import os
+from io import BytesIO
+from datetime import datetime
+from urllib.parse import urljoin
+from functools import lru_cache, reduce
+from typing import List, Dict, NamedTuple, Optional
+
+import pandas as pd
+import pytz
+
+from resdk import Resolwe
+from resdk.resources import Collection
+from resdk.utils.table_cache import default_cache_dir, load_cache, save_cache
+
+
+TPM = 'tpm'
+COUNTS = 'rc'
+EXPRESSION_TYPES = [TPM, COUNTS]
+
+META = 'meta'
+
+CHUNK_SIZE = 1000
+
+
+class TableSample(NamedTuple):
+    sample_slug: str
+    expression_files: Dict[str, str]
+    metadata: dict
+
+
+class CollectionTables:
+    """
+    A helper class to fetch given collection's expression and meta data in a tables format. All returned tables
+    have samples in rows and gene symbols in columns.
+
+    When calling :meth:`CollectionTables.expressions`, :meth:`CollectionTables.counts` and
+    :meth:`CollectionTables.metadata` for the first time the corresponding data gets fetched from the server.
+    This data than gets cached in memory and on disc and is used in consequent calls until the modified
+    attribute change.
+    """
+
+    def __init__(self, resolwe_obj: Resolwe, collection_slug: str, cache_dir=None):
+        """
+        :param resolwe_obj: initialized Resolwe object
+        :param collection_slug: slug of collection of interest
+        :param cache_dir: cache directory location, if not specified system specific cache directory is used
+        """
+        self.resolwe_obj = resolwe_obj
+        self.collection = self.resolwe_obj.collection.get(slug=collection_slug)  # type: Collection
+
+        self.cache_dir = cache_dir
+        if self.cache_dir is None:
+            self.cache_dir = default_cache_dir()
+        if not os.path.exists(self.cache_dir):
+            os.makedirs(self.cache_dir)
+
+        self._table_samples_ = []  # type: List[TableSample]
+        self._meta = None  # type: Optional[pd.DataFrame]
+        self._expressions = {}  # type: Dict[str, pd.DataFrame]
+        self._gene_map_ = {}  # type: Dict[str, str]
+        self._modified = None  # type: Optional[datetime]
+
+    @lru_cache(maxsize=16)
+    def expressions(self, preprocess=None) -> pd.DataFrame:
+        """Return TPM data in a table format.
+
+        If a data preprocess is set then its executed and its results get cached in memory.
+
+        :param preprocess: a callable that preprocess expression data
```

Comment:
Maybe add that the callable is called with a single argument of type `pd.DataFrame`.